### ISSUE TITLE
jdk15+36: java/foreign/TestMismatch.java.TestMismatch J9 Crash

### ISSUE THREAD

**Issue by andrew-m-leonard** [eclipse/openj9#10588]
Failure link
------------
https://ci.adoptopenjdk.net/job/Test_openjdk15_j9_sanity.openjdk_ppc64le_linux/43/consoleFull
java/foreign/TestMismatch.java
Fails on platforms: Platforms: pLinux, xLinuxXL

Optional info
-------------
- intermittent failure (yes|no): possibly
- regression or new test:  Regression
- if regression, what are the last passing / first failing public SHAs (OpenJ9, OMR, JCL) :

Failure output (captured from console output)
---------------------------------------------
```
14:13:28  testDifferentValues s1:MemorySegment{ id=0x6c3ff368 limit: 1 }, s2:MemorySegment{ id=0x266c2bbe limit: 1 }
14:13:28  test TestMismatch.testDifferentValues(jdk.internal.foreign.HeapMemorySegmentImpl@6bd39stderr:
14:13:28  Unhandled exception
14:13:28  Type=Segmentation error vmState=0x00000000
14:13:28  J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
14:13:28  Handler1=0000794DBD6EA980 Handler2=0000794DBD5479A0
14:13:28  R0=0000000080000009 R1=0000794D9F01AC50 R2=0000000000000000 R3=0000794D36000000
14:13:28  R4=FFFFFFFFFC000000 R5=0000000000000000 R6=0000794DBE4701C8 R7=0000000000000000
14:13:28  R8=0000000000000001 R9=0000000000000020 R10=0000000000000000 R11=FFFFFFFFFFFFF001
14:13:28  R12=0000000000000000 R13=0000794D9F0268E0 R14=0000000000324DC0 R15=000000000031F900
14:13:28  R16=0000794D9D020038 R17=FFFFFFFFFFFFFFFF R18=0000794D9F01AE70 R19=0000000000000000
14:13:28  R20=0000000000324F90 R21=00000000002C0000 R22=000000000031F900 R23=0000000000000001
14:13:28  R24=0000000080000009 R25=0000000000000000 R26=0000794CB5FF0040 R27=0000000080000009
14:13:28  R28=0000000000055160 R29=00000000002C0000 R30=0000000080000009 R31=0000000000000001
14:13:28  NIP=0000794D9F771FD4 MSR=800000000280D033 ORIG_GPR3=C000000000337DB4 CTR=FFFFFFFFF7FFF802
14:13:28  LINK=0000794D9F771F84 XER=0000000020000000 CCR=0000000048004282 SOFTE=0000000000000001
14:13:28  TRAP=0000000000000300 DAR=0000794D36000000 dsisr=0000000042000000 RESULT=0000000000000000
14:13:28  FPR0 0000794d3d8b09b9 (f: 1032522176.000000, d: 6.589485e-310)
14:13:28  FPR1 40452f4e40000000 (f: 1073741824.000000, d: 4.236958e+01)
14:13:28  FPR2 4059000000000000 (f: 0.000000, d: 1.000000e+02)
14:13:28  FPR3 3fee666660000000 (f: 1610612736.000000, d: 9.500000e-01)
14:13:28  FPR4 3fce840b4ac4e4d2 (f: 1254417664.000000, d: 2.384047e-01)
14:13:28  FPR5 bfe7154748bef6c8 (f: 1220474624.000000, d: -7.213475e-01)
14:13:28  FPR6 3fe62e42fefa39ef (f: 4277811712.000000, d: 6.931472e-01)
14:13:28  FPR7 3fbc5e53aa362eb4 (f: 2855677696.000000, d: 1.108143e-01)
14:13:28  FPR8 006f0076006e0049 (f: 7209033.000000, d: 1.379626e-306)
14:13:28  FPR9 002000720065006b (f: 6619243.000000, d: 4.450632e-308)
14:13:28  FPR10 3809c01c80000000 (f: 2147483648.000000, d: 9.459216e-39)
14:13:28  FPR11 0000000000000040 (f: 64.000000, d: 3.162020e-322)
14:13:28  FPR12 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR13 0000000000000001 (f: 1.000000, d: 4.940656e-324)
14:13:28  FPR14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR16 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR17 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR18 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR19 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR20 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR21 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR22 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR23 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR24 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR25 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR26 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR27 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR28 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR29 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR30 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  FPR31 0000000000000000 (f: 0.000000, d: 0.000000e+00)
14:13:28  
14:13:28  Compiled_method=jdk/internal/foreign/NativeMemorySegmentImpl.makeNativeSegment(JJ)Ljdk/incubator/foreign/MemorySegment;
14:13:28  Target=2_90_20200911_34 (Linux 4.15.0-45-generic)
14:13:28  CPU=ppc64le (4 logical CPUs) (0x1fe0c0000 RAM)
14:13:28  ----------- Stack Backtrace -----------
14:13:28  (0x0000794D9F771FD4 [<unknown>+0x0])
14:13:28  (0x0000794DBD6D23AC [libj9vm29.so+0x923ac])
14:13:28  (0x0000794DBD6D36D8 [libj9vm29.so+0x936d8])
14:13:28  (0x0000794DBC4E55C4 [libjclse29.so+0x755c4])
14:13:28  JVM_InvokeMethod+0x40 (0x0000794DBD872970 [libjvm.so+0x22970])
14:13:28  JVM_InvokeMethod+0x34 (0x0000794DBE216884 [libjvm.so+0x6884])
14:13:28  Java_jdk_internal_reflect_NativeMethodAccessorImpl_invoke0+0x24 (0x0000794D9CC9BFA4 [libjava.so+0xbfa4])
14:13:28  (0x0000794D9F4F59A4 [<unknown>+0x0])
14:13:28  (0x0000794DBD6D23AC [libj9vm29.so+0x923ac])
14:13:28  (0x0000794DBD6D36D8 [libj9vm29.so+0x936d8])
14:13:28  (0x0000794DBC4E55C4 [libjclse29.so+0x755c4])
14:13:28  JVM_InvokeMethod+0x40 (0x0000794DBD872970 [libjvm.so+0x22970])
14:13:28  JVM_InvokeMethod+0x34 (0x0000794DBE216884 [libjvm.so+0x6884])
14:13:28  Java_jdk_internal_reflect_NativeMethodAccessorImpl_invoke0+0x24 (0x0000794D9CC9BFA4 [libjava.so+0xbfa4])
14:13:28  (0x0000794D9F4F59A4 [<unknown>+0x0])
14:13:28  (0x0000794DBD6CD6B4 [libj9vm29.so+0x8d6b4])
14:13:28  (0x0000794DBD740C30 [libj9vm29.so+0x100c30])
14:13:28  (0x0000794DBD548DA8 [libj9prt29.so+0x28da8])
14:13:28  (0x0000794DBD73C264 [libj9vm29.so+0xfc264])
14:13:28  (0x0000794DBD611248 [libj9thr29.so+0x11248])
14:13:28  (0x0000794DBE4B885C [libpthread.so.0+0x885c])
14:13:28  clone+0x98 (0x0000794DBE3A9028 [libc.so.6+0x159028])
14:13:28  ---------------------------------------

```

**Comment by JasonFengJ9** [comment]
It appears https://github.com/eclipse/openj9/issues/10580 with different error output.

**Comment by DanHeidinga** [comment]
> Compiled_method=jdk/internal/foreign/NativeMemorySegmentImpl.makeNativeSegment(JJ)Ljdk/incubator/foreign/MemorySegment;

fyi @andrewcraik

**Comment by DanHeidinga** [comment]
We need a stop-ship determination on this issue.  Given the crash is in java.foreign package which I believe is still in preview, I don't think will be but we should still look at the crash to be sure it's a broader issue

**Comment by andrewcraik** [comment]
@liqunl sorry to interrupt your dev work but could you take a look as this is lined up against the release that will be going out imminently.

**Comment by liqunl** [comment]
@andrew-m-leonard @llxia What should be the target if I want to rerun the failed test?

**Comment by liqunl** [comment]
@andrew-m-leonard You mentioned the test also failed on xlinux, could you provide the link to the x86 failure?

**Comment by liqunl** [comment]
#10580 is a failure in the same test and it has a link to rerun the test with target `jdk_foreign_0`. 
20 iterations
https://ci.adoptopenjdk.net/job/Grinder/3856/

But it failed. @rpshukla Could you help with the grinder? I don't know why my grinder failed, maybe something's wrong in my configuration?

**Comment by andrew-m-leonard** [comment]
xLinux: https://ci.adoptopenjdk.net/job/Test_openjdk15_j9_sanity.openjdk_x86-64_linux_xl/41/console

**Comment by andrew-m-leonard** [comment]
@liqunl i've just kicked off a Grinder on xLinux for just the TestMissmatch test here: https://ci.adoptopenjdk.net/job/Grinder/3859/console

**Comment by andrew-m-leonard** [comment]
@liqunl fyi, the "Rerun Grinder" links rarely work as-is, as they typically point at the "upstream" job artifact, which by the time you run the grinder has dissappeared!
I change it to use "customized" and point SDK_URL at the required github binary

**Comment by rpshukla** [comment]
I started another plinux grinder using CUSTOMIZED_SDK_URL instead of UPSTREAM_JOB_NAME
https://ci.adoptopenjdk.net/job/Grinder/3861/

**Comment by liqunl** [comment]
Looked at the xlinuxXL failure, the error message is the same as #10580, which is 
```
TEST RESULT: Error. Agent error: java.lang.Exception: Agent 2 timed out with a timeout of 960 seconds; check console log for any additional details
```
So the crash is only seen on ppcle. FYI @gita-omr

**Comment by gita-omr** [comment]
@AlenBadel could you kindly take a look?

**Comment by AlenBadel** [comment]
Failure rate on my side is between 1/10 and 1/20. 

>  Given the crash is in java.foreign package which I believe is still in preview, I don't think will be but we should still look at the crash to be sure it's a broader issue

java.foreign was bundled with OpenJDK and Oracle builds starting with Java 14. I understand J9 may not be in total sync with new package support (i.e javafx). Is foreign bundled with J9 Java 14 builds?

**Comment by pshipton** [comment]
Foreign is still incubator status for Java 15, but it is included.
https://openjdk.java.net/jeps/383

**Comment by AlenBadel** [comment]
It looks like the foreign package is also included in J9 Java 14 builds. I ran it with the latest Java 14 release, and was not able to reproduce the failure. This suggests that this is a recent regression.
I'm still actively attempting to open the core file, and analyzing the failure.

**Comment by AlenBadel** [comment]
I'm afraid that the core files are not readable on my end. I've even tried to add `-Xrs` in order to get more core context and that causes EOF issues with the test, none of the iterations have crashed using `-Xrs`.

Continuing my investigation by attempting to increase the reproducibility, so that I can reliably get this test to fail locally.

**Comment by pshipton** [comment]
I'm guessing the core file(s) aren't readable because you need the jextract data from the machine that produced the core file. If it's reproducible on an OpenJ9 or internal jenkins machine we can get that. If it's only reproducible at Adopt, someone with access to those machines could run jextract.

**Comment by DanHeidinga** [comment]
This is considered stop-ship until we have a core file and can evaluate the cause and frequency

**Comment by liqunl** [comment]
It's reproducible on internal jenkins machine, failure rate seems to be 4/12 on nightly build. Also saw crash in compiled method `jdk/internal/misc/Unsafe.setMemory(JJB)V`
The following grinder is to reproduce the timeout, and it hasn't finished yet. 
https://hyc-runtimes-jenkins.swg-devops.com/job/Grinder/10355

**Comment by AlenBadel** [comment]
I was able to re-produce this locally on one of our internal farm machines. Failing at a rate of 100% on this particular machine. 

jextract can't find the libraries unfortunately. GDB, as well as our internal debugging tools are not able to load the libraries even when given the prefix, and absolute path. 
```
/workspace/ws/openjdk-tests$ ../jdk-15+36/bin/jextract ./TKG/test_output_16002861801429/jdk_foreign_0/work/java/foreign/TestSegments/core.20200916.155655.9347.0001.dmp
Loading dump file...
Read memory image from ./TKG/test_output_16002861801429/jdk_foreign_0/work/java/foreign/TestSegments/core.20200916.155655.9347.0001.dmp
Creating archive file: ./TKG/test_output_16002861801429/jdk_foreign_0/work/java/foreign/TestSegments/core.20200916.155655.9347.0001.dmp.zip
Adding "./TKG/test_output_16002861801429/jdk_foreign_0/work/java/foreign/TestSegments/core.20200916.155655.9347.0001.dmp"
Warning:  Could not find file "libgcc_s.so.1" for inclusion in archive
Warning:  Could not find file "libnet.so" for inclusion in archive
Warning:  Could not find file "libjava.so" for inclusion in archive
Warning:  Could not find file "ld64.so.2" for inclusion in archive
Warning:  Could not find file "libzip.so" for inclusion in archive
Warning:  Could not find file "librt.so.1" for inclusion in archive
Warning:  Could not find file "libm.so.6" for inclusion in archive
Warning:  Could not find file "libz.so.1" for inclusion in archive
Warning:  Could not find file "libjclse29.so" for inclusion in archive
Warning:  Could not find file "libjncrypto.so" for inclusion in archive
Warning:  Could not find file "libj9zlib29.so" for inclusion in archive
Warning:  Could not find file "libdl.so.2" for inclusion in archive
Warning:  Could not find file "libpthread.so.0" for inclusion in archive
Warning:  Could not find file "libnss_nis.so.2" for inclusion in archive
Warning:  Could not find file "libj9dmp29.so" for inclusion in archive
Warning:  Could not find file "libj9trc29.so" for inclusion in archive
Warning:  Could not find file "libc.so.6" for inclusion in archive
Warning:  Could not find file "libjimage.so" for inclusion in archive
Warning:  Could not find file "libjvm.so" for inclusion in archive
Warning:  Could not find file "linux-vdso64.so.1" for inclusion in archive
Warning:  Could not find file "libj9vm29.so" for inclusion in archive
Warning:  Could not find file "libjli.so" for inclusion in archive
Warning:  Could not find file "libj9prt29.so" for inclusion in archive
Warning:  Could not find file "libnss_compat.so.2" for inclusion in archive
Adding "/home/jenkins/workspace/ws/jdk-15+36/bin/java"
Warning:  Could not find file "libomrsig.so" for inclusion in archive
Warning:  Could not find file "libj9hookable29.so" for inclusion in archive
Warning:  Could not find file "libj9vrb29.so" for inclusion in archive
Warning:  Could not find file "libnsl.so.1" for inclusion in archive
Warning:  Could not find file "libj9thr29.so" for inclusion in archive
Warning:  Could not find file "libj9gc29.so" for inclusion in archive
Warning:  Could not find file "libj9shr29.so" for inclusion in archive
Warning:  Could not find file "libnio.so" for inclusion in archive
Warning:  Could not find file "libnss_files.so.2" for inclusion in archive
Warning:  Could not find file "libj9jvmti29.so" for inclusion in archive
Warning:  Could not find file "libextnet.so" for inclusion in archive
Adding "/home/jenkins/workspace/ws/jdk-15+36/lib/J9TraceFormat.dat"
Adding "/home/jenkins/workspace/ws/jdk-15+36/lib/OMRTraceFormat.dat"
jextract complete.
```

I've also tried to generate a trace file, without any luck. 

I'll attempt to run this workload with gdb to get a snapshot of where we're crashing.

**Comment by pshipton** [comment]
The jextract problem may be related to https://github.com/eclipse/openj9/issues/9549, for which we have a PR under review. @keithc-ca 

Although it makes no sense that GDB doesn't work. Perhaps it needs to be updated.

**Comment by AlenBadel** [comment]
> The jextract problem may be related to #9549, for which we have a PR under review. @keithc-ca

Thanks. I'll give the build a try to see if it addresses my issue. 
 
> Although it makes no sense that GDB doesn't work. Perhaps it needs to be updated.
I'm currently using GDB 8.2, which is older. However it's the most current available build given the OS I'm using is ubuntu 16.04. 

I'm just shooting in the dark here, but I wonder if I can get this to reproduce on a newer OS, there might be a bug with some of these older c libraries and runtimes; specially on PPC64LE.

**Comment by keithc-ca** [comment]
I vaguely remember having trouble with gdb 8 and needing to use gdb 7. There's a copy for x86-64 in /team/keithc/gdb-7, but, unfortunately, I don't know where to find gdb 7 for ppc64le.

**Comment by AlenBadel** [comment]
In a turn of events. This OpenJ9 JDK15 nightly is now starting to hang during a simple `java --version` on the same machine.

**Comment by AlenBadel** [comment]
Looks to be hanging only on Non XL releases of PPC64LE JDK15. Works on the JDK14 equivalent. I went back all the way to the first release of ppc64le JDK15 on adoptopenjdk, and they all hang similarly.

**Comment by keithc-ca** [comment]
Which nightly build are you using? I don't have that problem with the latest nightly from adoptopenjdk.net:
```
$ ./bootjdk15/bin/java -version
openjdk version "15" 2020-09-15
OpenJDK Runtime Environment AdoptOpenJDK (build 15+36-202009162112)
Eclipse OpenJ9 VM AdoptOpenJDK (build master-35e1cad7f, JRE 15 Linux ppc64le-64-Bit Compressed References 20200916_41 (JIT enabled, AOT enabled)
OpenJ9   - 35e1cad7f
OMR      - af16e8970
JCL      - d331dd5764b based on jdk-15+36)
```
And using that as the boot jdk, I was able to produce a local build:
```
$ ./images/jdk/bin/java -version
openjdk version "15-internal" 2020-09-15
OpenJDK Runtime Environment (build 15-internal+0-adhoc.keithc.jdk15)
Eclipse OpenJ9 VM (build main-a3acbbf, JRE 15 Linux ppc64le-64-Bit Compressed References 20200917_000000 (JIT enabled, AOT enabled)
OpenJ9   - a3acbbf
OMR      - af16e89
JCL      - 09ea3d3 based on jdk-15+36)
```

**Comment by pshipton** [comment]
@AlenBadel there is a grinder in https://github.com/eclipse/openj9/issues/10580#issuecomment-694222270 which crashed 3/20 on a rhel7 machine. Does that help? Perhaps these cores will be readable or gdb is working on rhel7.

**Comment by AlenBadel** [comment]
I was able to get a stack trace for this crash. Investigating further.
```
#0  0x00003fff81bdffac in ?? ()
#1  0x00003fff80d51610 in omrdump_create () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9prt29.so
#2  0x00003fff80adcf34 in doSystemDump () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9dmp29.so
#3  0x00003fff80ad5c80 in protectedDumpFunction () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9dmp29.so
#4  0x00003fff80d049c8 in omrsig_protect () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9prt29.so
#5  0x00003fff80adbe80 in runDumpFunction () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9dmp29.so
#6  0x00003fff80adc080 in runDumpAgent () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9dmp29.so
#7  0x00003fff80afcc78 in triggerDumpAgents () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9dmp29.so
#8  0x00003fff80afa85c in rasDumpHookExceptionSysthrow () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9dmp29.so
#9  0x00003fff80d91840 in J9HookDispatch () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9hookable29.so
#10 0x00003fff80ea5b28 in setCurrentExceptionWithUtfCause () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#11 0x00003fff80ea605c in setCurrentExceptionUTF () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#12 0x00003fff80ea85b8 in setNativeOutOfMemoryError () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#13 0x00003fff80428140 in unsafeAllocateMemory () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjclse29.so
#14 0x00003fff80419570 in Java_sun_misc_Unsafe_allocateMemory () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjclse29.so
#15 0x00003fff5cb7cd60 in ?? ()
#16 0x00003fff80e929ec in sidecarInvokeReflectMethodImpl () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#17 0x00003fff80e93d18 in sidecarInvokeReflectMethod () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#18 0x00003fff80435204 in JVM_InvokeMethod_Impl () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjclse29.so
#19 0x00003fff810327c0 in JVM_InvokeMethod () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjvm.so
#20 0x00003fff81976884 in addXserviceArgs () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjvm.so
#21 0x00003fff562cbfa4 in ?? ()
#22 0x00003fff5cb35fa4 in ?? ()
#23 0x00003fff80e929ec in sidecarInvokeReflectMethodImpl () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#24 0x00003fff80e93d18 in sidecarInvokeReflectMethod () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#25 0x00003fff80435204 in JVM_InvokeMethod_Impl () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjclse29.so
#26 0x00003fff810327c0 in JVM_InvokeMethod () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjvm.so
#27 0x00003fff81976884 in addXserviceArgs () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libjvm.so
#28 0x00003fff562cbfa4 in ?? ()
#29 0x00003fff5cb35fa4 in ?? ()
#30 0x00003fff80e8dcf4 in runJavaThread () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#31 0x00003fff80f014d0 in javaProtectedThreadProc () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9vm29.so
#32 0x00003fff80d049c8 in omrsig_protect () from /home/jenkins/ws/jdk-15+36/lib/compressedrefs/libj9prt29.so
```

**Comment by AlenBadel** [comment]
The stack trace above is certainly not related to the crash within the compiled method
`jdk/internal/foreign/NativeMemorySegmentImpl.makeNativeSegment(JJ)Ljdk/incubator/foreign/MemorySegment`. 

We're seeing this mainly on systems that are resource starved. As is I have not been able to reproduce this on any baremetal machines or Power VMs with more than 8GB of Ram. 

However, running with `-Xjit:count=0` will allow the test to fail with certainty on almost any Power machine. 
Obtaining a core file has been quite difficult. The core file that is produced is never saved on disk, no matter what machine I've ran it on. I've also verified the ulimit values to be acceptable to produce a core file.  GDB also does not break upon segmentation.

Currently, I'm working on isolating and producing a trace of the compiled method we're crashing in, once that is done I hope to have a genuine core and stack trace.

**Comment by andrewcraik** [comment]
I would note `-Xjit:count=0` is a bit of a weird stress mode because nothing will be resolved at compile-time. It also forces the JIT to do synchronous compilation. It would be good to know if the synchronous compilation that is important or the super low count and does it also fail with `Xjit:count=1,disableAsyncCompilation`?

**Comment by AlenBadel** [comment]
> . It would be good to know if the synchronous compilation that is important or the super low count and does it also fail with `Xjit:count=1,disableAsyncCompilation`?

On the machine with ample resources, the crash does not appear with `Xjit:count=1,disableAsyncCompilation`. I'm attempting the same test on one of the machines that produces the failure under normal circumstances. The state of the machine I was using has changed, so I'll be running a Grinder to verify the result on other machines.

**Comment by gita-omr** [comment]
If count=0 makes it fail in exactly the same way as described when the defect was opened, e.g. crash in:

Compiled_method=jdk/internal/foreign/NativeMemorySegmentImpl.makeNativeSegment(JJ)Ljdk/incubator/foreign/MemorySegment;

then I think it would be ok to stick to count=0 and try to debug that. As I understand @AlenBadel managed to make it fail consistently that way. And this a native method btw.

**Comment by gita-omr** [comment]
Unless we are trying to determine the severity of this defect...

**Comment by andrewcraik** [comment]
I think the above was both to verify if the crash was specific to count=0 and to assess severity.

Looking at the most recent backtrace from Alen I see: `setNativeOutOfMemoryError` under the Unsafe API - seems like we are failing to allocate memory? Is this just a resource exhaustion issue? Is the test running single threaded or are there multiple things running on the machine? That kind of error could be something like being unable to grow the stack or unable to satisfy a native memory request...

**Comment by AlenBadel** [comment]
>  seems like we are failing to allocate memory? Is this just a resource exhaustion issue?

Yes, as I've mentioned I've only been able to reproduce this bug under normal conditions on a system with very constrained resources. We're talking about 8GB provisioned PowerVMs on a heavily cached system. I have not been able to reproduce this upon restarting and dropping these caches, as well as on any larger systems. I.e a 32GB PowerVM or a baremetal.

>  Is the test running single threaded or are there multiple things running on the machine? 

The tests for the most part are single threaded. At most it will create a worker thread to conduct a test and waits for it to finish.

**Comment by gita-omr** [comment]
> However, running with `-Xjit:count=0` will allow the test to fail with certainty on almost any Power machine.

Does it fail with the same setNativeOutOfMemoryError issue?

**Comment by gita-omr** [comment]
In any case, I guess we should look into how to exist gracefully from setNativeOutOfMemoryError when resources are limited.

**Comment by pshipton** [comment]
After a discussion in Slack, this has been determined not a stop ship issue. The 0.22 release has been finalized and this issue is moved to the 0.23 release for which we've already branched. The work to fix it should continue at the same priority so we have a fix by Oct 2 for 0.23 M2.

**Comment by AlenBadel** [comment]
> Does it fail with the same setNativeOutOfMemoryError issue?

To summarize:

**Running the test as-is (Without Overriding JVM Options) on a resource constraint system**
- The test will fail with the segmentation fault we're seeing. The core file is not saved on disk.
- An additional core file may be saved to disk with the `setNativeOutOfMemoryError`, this core is not a dump of the thread that encountered the segmentation fault. I'm unsure of the origins of this core. 

**Running the test as-is on a system with ample resources** 
- The test passes. No core files produced.

**Running the test with `-Xjit:count=0` on a system with ample resources**
- A segmentation fault in the same compiled method occurs. The core file is not saved on disk.
- No additional core files produced.

**Comment by gita-omr** [comment]
Thanks for the summary. I would propose to focus on **Running the test with -Xjit:count=0 on a system with ample resources** but I am open to other suggestions.

**Comment by pshipton** [comment]
https://ci.eclipse.org/openj9/job/Test_openjdk15_j9_sanity.openjdk_ppc64le_linux_Nightly/55
```
02:20:31  testDifferentValues s1:MemorySegment{ id=0x47d2738 limit: 16 }, s2:MemorySegment{ id=0x47c6f28 limit: 16 }
02:20:31  test TestMismatch.testDifferentValues(jdk.internal.foreign.NativeMemorySegmentImpl@c288db05, jdk.internal.foreign.NativeMemorySegmentImpl@13bb5919): success
02:20:31  testDifferentValues s1:MemorySegment{ id=0x47d2738 limit: 16 }, s2:MemorySegment{ id=0x47c6f28 limit: 8 }
02:20:31  test TestMismatch.testDifferentValues(jdk.internal.foreign.NativeMemorySegmentImpl@c288db05, jdk.internal.foreign.NativeMemorySegmentImpl@125597be): success
02:20:31  testDifferentValues s1:MemorySegment{ id=0x47d2738 limit: 16 }, s2:MemorySegment{ id=0x47c6e30 limit: 8 }
...
02:20:31  test TestMismatch.testDifferentValues(jdk.internal.foreign.HeapMemorySegmentImpl@98eb0d2a, jdk.internal.foreign.HeapMemorySegmentImpl@37d37b94): success
02:20:31  testDifferentValues s1:MemorySegment{ id=0x609cd924 limit: 1 }, s2:MemorySegment{ id=0x17282262 limit: stderr:
02:20:31  Unhandled exception
02:20:31  Type=Segmentation error vmState=0x00000000
02:20:31  J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000002
02:20:31  Handler1=00003FFFB4A6C2A0 Handler2=00003FFFB48C35C0
02:20:31  
02:20:31  Unhandled exception in signal handler. Protected function: writeCrashDataToConsole (0x0)
02:20:31  
02:20:31  ----------- Stack Backtrace -----------
02:20:31  Unhandled exception
02:20:31  Type=Segmentation error vmState=0x00000000
02:20:31  J9Generic_Signal_Number=00000018 Signal_Number=0000000b Error_Value=00000000 Signal_Code=00000001
02:20:31  Handler1=00003FFFB4A6C2A0 Handler2=00003FFFB48C35C0
02:20:31  R0=00003FFFB4B519BC R1=00003FFF94139F60 R2=00003FFFB4B96F00 R3=00003FFFB0157410
02:20:31  R4=00003FFF9413A1D8 R5=0000000000000000 R6=FFFFFFFFFFFFFFFF R7=00003FFF5D4619D3
02:20:31  R8=0000000000000000 R9=0000000000000000 R10=0000000000000724 R11=0000000000000000
02:20:31  R12=0000000000002200 R13=00003FFF94146900 R14=00000000002D59D0 R15=00000000001B2200
02:20:31  R16=00003FFFB4B92EC8 R17=0000000000000000 R18=00003FFFB0098DC8 R19=0000000000000000
02:20:31  R20=0000000000004412 R21=0000000000000000 R22=00003FFF9413A688 R23=00003FFFB0157428
02:20:31  R24=FFFFFFFFFFFFFA08 R25=0000000000000003 R26=00003FFFB0156E40 R27=00003FFFB4B97208
02:20:31  R28=00003FFFB0156E30 R29=00003FFF9413A1D8 R30=00003FFFB0157410 R31=00003FFF280C8720
02:20:31  NIP=00003FFFB4B5563C MSR=800000000280F033 ORIG_GPR3=00003FFFAF0B3B88 CTR=000000000000000A
02:20:31  LINK=00003FFFB4B519BC XER=0000000020000000 CCR=0000000048222222 SOFTE=0000000000000001
02:20:31  TRAP=0000000000000300 DAR=0000000000000000 dsisr=0000000040000000 RESULT=0000000000000000
02:20:31  FPR0 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR1 4055413e00000000 (f: 0.000000, d: 8.501941e+01)
02:20:31  FPR2 00003fff94138250 (f: 2484306432.000000, d: 3.476588e-310)
02:20:31  FPR3 3fc7466496cb03de (f: 2529887232.000000, d: 1.818357e-01)
02:20:31  FPR4 3fcc71c51d8e78af (f: 495876256.000000, d: 2.222220e-01)
02:20:31  FPR5 3fd2492494229359 (f: 2485293824.000000, d: 2.857143e-01)
02:20:31  FPR6 00003fffb4c31e50 (f: 3032686080.000000, d: 3.476616e-310)
02:20:31  FPR7 00003fffb4c31e50 (f: 3032686080.000000, d: 3.476616e-310)
02:20:31  FPR8 3700b24d0200bd1b (f: 33602844.000000, d: 9.358705e-44)
02:20:31  FPR9 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR10 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR11 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR12 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR13 0000000000000001 (f: 1.000000, d: 4.940656e-324)
02:20:31  FPR14 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR15 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR16 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR17 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR18 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR19 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR20 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR21 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR22 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR23 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR24 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR25 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR26 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR27 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR28 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR29 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR30 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  FPR31 0000000000000000 (f: 0.000000, d: 0.000000e+00)
02:20:31  Module=/home/jenkins/jenkins-agent/workspace/Test_openjdk15_j9_sanity.openjdk_ppc64le_linux_Nightly/openjdkbinary/j2sdk-image/lib/compressedrefs/libj9vm29.so
02:20:31  Module_base_address=00003FFFB49C0000
02:20:31  Target=2_90_20200923_57 (Linux 4.4.0-189-generic)
02:20:31  CPU=ppc64le (4 logical CPUs) (0x1fe380000 RAM)
02:20:31  ----------- Stack Backtrace -----------
02:20:31  (0x00003FFFB4B5563C [libj9vm29.so+0x19563c])
02:20:31  (0x000000005F6AE476 [<unknown>+0x0])
02:20:31  (0x00003FFFB4B519BC [libj9vm29.so+0x1919bc])
02:20:31  (0x00003FFFB4B527A0 [libj9vm29.so+0x1927a0])
02:20:31  (0x00003FFFB4A2FE2C [libj9vm29.so+0x6fe2c])
02:20:31  (0x00003FFFB4A550E8 [libj9vm29.so+0x950e8])
02:20:31  (0x00003FFFB4A62D98 [libj9vm29.so+0xa2d98])
02:20:31  (0x00003FFFB4AFC0CC [libj9vm29.so+0x13c0cc])
02:20:31  (0x00003FFFB4A58A18 [libj9vm29.so+0x98a18])
02:20:31  (0x00003FFFB4A5A69C [libj9vm29.so+0x9a69c])
02:20:31  (0x00003FFFB4A5C778 [libj9vm29.so+0x9c778])
02:20:31  (0x00003FFFB49DE45C [libj9vm29.so+0x1e45c])
02:20:31  (0x00003FFFB4AA0A0C [libj9vm29.so+0xe0a0c])
02:20:31  (0x00003FFFB4A4CE0C [libj9vm29.so+0x8ce0c])
02:20:31  (0x00003FFFB4A5B2D8 [libj9vm29.so+0x9b2d8])
02:20:31  (0x00003FFFB4A5C8AC [libj9vm29.so+0x9c8ac])
02:20:31  (0x00003FFFB4AA69F8 [libj9vm29.so+0xe69f8])
02:20:31  (0x00003FFFB4AA7094 [libj9vm29.so+0xe7094])
02:20:31  (0x00003FFFB4AA7434 [libj9vm29.so+0xe7434])
02:20:31  (0x00003FFFB49E19E0 [libj9vm29.so+0x219e0])
02:20:31  (0x00003FFFB4AA0A0C [libj9vm29.so+0xe0a0c])
02:20:31  (0x00003FFFB4A4EEB4 [libj9vm29.so+0x8eeb4])
02:20:31  (0x00003FFFB4AC1B30 [libj9vm29.so+0x101b30])
02:20:31  (0x00003FFFB48C49C8 [libj9prt29.so+0x349c8])
02:20:31  (0x00003FFFB4ABD144 [libj9vm29.so+0xfd144])
02:20:31  (0x00003FFFB4991248 [libj9thr29.so+0x11248])
02:20:31  (0x00003FFFB5778040 [libpthread.so.0+0x8040])
02:20:31  clone+0x98 (0x00003FFFB5694230 [libc.so.6+0x124230])
02:20:31  ---------------------------------------

```

**Comment by keithc-ca** [comment]
The latest crash is worse: the handler crashed:
```
02:20:31  Unhandled exception in signal handler. Protected function: writeCrashDataToConsole (0x0)
```

**Comment by AlenBadel** [comment]
> The latest crash is worse: the handler crashed:
> 
> ```
> 02:20:31  Unhandled exception in signal handler. Protected function: writeCrashDataToConsole (0x0)
> ```

Right. We've been seeing that all along, which is why the core files never end up being saved to disk.

**Comment by pshipton** [comment]
> which is why the core files never end up being saved to disk

There is a core file, what I assume you are saying is that the core file is for the writeCrashDataToConsole issue and not the original issue. @keithc-ca can you please help look at the writeCrashDataToConsole problem.

**Comment by AlenBadel** [comment]
I was able to narrow the problem down to this simple test case.
```
import jdk.incubator.foreign.MemorySegment;

public class TestMemorySegmentAllocateNative {
    public static void main(String args[]) {
        var s1 = MemorySegment.allocateNative((long)Integer.MAX_VALUE + 10L);
    }
}
```

I'm able to get a core file via GDB, however a normal invocation of this test cases the handler to hang attempting to create the core file.

**Comment by gita-omr** [comment]
What happens with -Xint?

**Comment by AlenBadel** [comment]
> What happens with -Xint?

Test Passes with `-Xint`, and passes allocating smaller memory segments.

**Comment by andrewcraik** [comment]
Is the `-Xint` pass because we run less allocations total or do we do the same allocation with the JIT it just isn't satisfied or something?

**Comment by AlenBadel** [comment]
> Is the -Xint pass because we run less allocations total or do we do the same allocation with the JIT it just isn't satisfied or something?

We're running the same test with the same number of allocations, both with and without -Xint.

Updating with the stack trace. 
```
#0  0x00007fffcfb204a4 in {jdk/internal/foreign/NativeMemorySegmentImpl.makeNativeSegment} JIT ()
#1  0x00007ffff5f24d98 in startproc._interpreterUnresolvedInstanceDataGlue () from /root/build/builds/jdk-15+36/lib/compressedrefs/libj9jit29.so
#2  0x00007ffff70b32e8 in runCallInMethod (env=0x0, receiver=0x7ffff6fb1860 <J9HookDispatch(J9HookInterface**, uintptr_t, void*)+448>, clazz=0x7ffff7a9ded0,
    methodID=0x7ffe72ba8, args=0x7ffff6fd7f00) at callin.cpp:1083
#3  0x00007ffff70d1260 in gpProtectedRunCallInMethod (entryArg=0x7ffff7a9e638) at jnicsup.cpp:259
#4  0x00007ffff712f728 in signalProtectAndRunGlue (portLibrary=<optimized out>, userData=<optimized out>) at jniprotect.c:45
#5  0x00007ffff6f249c8 in omrsig_protect (portLibrary=0x7ffff7291e50 <j9portLibrary>, fn=0x7ffff712f700 <signalProtectAndRunGlue>, fn_arg=0x7ffff7a9e5d0,
    handler=0x7ffff70cc2a0 <structuredSignalHandler>, handler_arg=0x18200, flags=<optimized out>, result=0x7ffff7a9e5c8) at ../../omr/port/unix/omrsignal.c:425
#6  0x00007ffff712f7fc in gpProtectAndRun (function=0x7ffff70d1210 <gpProtectedRunCallInMethod(void*)>, env=0x18200, args=0x7ffff7a9e638) at jniprotect.c:78
#7  0x00007ffff70d3e74 in gpCheckCallin (env=0x18200, receiver=0x0, cls=0x103cc0, methodID=<optimized out>, args=<optimized out>) at jnicsup.cpp:447
#8  0x00007ffff70d0868 in callStaticVoidMethod (env=<optimized out>, cls=<optimized out>, methodID=<optimized out>) at jnicgen.c:288
```

**Comment by AlenBadel** [comment]
Narrowing it further, 

- This is reproduced by limiting compilation to `jdk/internal/foreign/NativeMemorySegmentImpl.makeNativeSegment` hence we don't need to worry about any callers being compiled.

- Crash only occurs when inlining is enabled. It's reproducible when inlining is the only optimization enabled.

- Further, allocating a Memory Segment larger than `Integer.MAX_VALUE` causes the failure. This suggests that there's an overflow issue.

**Comment by AlenBadel** [comment]
Precisely, the crash is occurring in the following portion of the compiled method. 

```
    0x7fffcfb204a4 0000042c [    0x7fffcec19e50]                   6    Label L0498:
    0x7fffcfb204a4 0000042c [    0x7fffcec19f90] f8430000          6    std     [gr3, 0], gr2
    0x7fffcfb204a8 00000430 [    0x7fffcec1a0e0] f8430008          6    std     [gr3, 8], gr2
    0x7fffcfb204ac 00000434 [    0x7fffcec1a230] f8430010          6    std     [gr3, 16], gr2
    0x7fffcfb204b0 00000438 [    0x7fffcec1a380] f8430018          6    std     [gr3, 24], gr2
    0x7fffcfb204b4 0000043c [    0x7fffcec1a420] 38630020          6    addi    gr3, gr3, 32
    0x7fffcfb204b8 00000440 [    0x7fffcec1a4c0] 4200ffec          6    bdnz    cr0, Label L0498
```

My assumption is that this loop is actively clearing memory allocated.

Specifically,
`r2` contains 0. 
 `r3` starts off with the head of the memory to be allocated, and is incremented to subsequent segments as each segment is cleared. 
The number of segments to be cleared is determined the value stored into the count register. 
The loop terminates when the counter reaches to 0.

Just before entering the loop the count register is set to `0xfffffffffc000000`. An invalid counter value, which could be due to overflow. This results in r3 reaching read-only memory, which causes the seg fault we're seeing. 

This is confirmed by looking at the javacore
`1XHEXCPCODE    Signal_Code: 00000002` 
`SEGV_ACCERR	2	/* invalid permissions for mapped object */`

As a side note. I suspect the value of the counter should be actually slightly less than 0x4001FFE. That value comes from the delta between the value of the counter initially to the moment it encounters the read-only memory. Which is about 2.15GB.

**Comment by AlenBadel** [comment]
The size of the segment inside the java stack is correct. This issue is stems from an instruction call srawi which only shifts the lower 32 bits.  I.e it sounds like somewhere we're assuming the generated size to be an integer. 

```
ld      gr0, [gr14, 176]                # SymRef  alignedSize<auto slot 
...
srawi   gr4, gr0, 5
mtctr   gr4
```

**Comment by AlenBadel** [comment]
Changing `srawi` to `sradi` looks to resolve the issue on Power.

https://github.com/AlenBadel/omr/commit/0b9653c23ba22247f66c11ad6df8a3b46c2c4738

I'll be taking a look on xLinuxXL to see if it's a similar story.

**Comment by gita-omr** [comment]
> it sounds like somewhere we're assuming the generated size to be an integer.

I think we need to find out  which opcode (or some recognized method?) this instruction is coming from and check/document overall assumptions about the incoming length for that code.

**Comment by IBMJimmyk** [comment]
From what Alen showed me, it comes from `Unsafe.setMemory(Ljava/lang/Object;JJB)V`. The third parameter is the size of the data to write and it accepts a Long value. The generated code treating it like an int value is wrong.

**Comment by AlenBadel** [comment]
Running the same test https://github.com/eclipse/openj9/issues/10588#issuecomment-698430817 on XLinuxXL does not produce a crash, or does it hang. 

I will continue attempting to reproduce this on XLinuxXL, but I believe the reproduction rate was established to be less than 1 in 300. https://github.com/eclipse/openj9/issues/10580#issuecomment-694222270

**Comment by andrewcraik** [comment]
@IBMJimmyk is a tree problem or a codegen problem just to be clear on the width issue at play here? I'm guessing codegen?

**Comment by IBMJimmyk** [comment]
At least the problem on Power was a codegen issue. The child for the size of the data to write was an `lload` which is expected. What was going on was the call to `Unsafe.setMemory` was being recognized as `sun_misc_Unsafe_setMemory`.  Part way into generating code for a recognized call to `Unsafe.setMemory` the wrong data size is used from an instruction which treated the data as if it was int and not long. That's what caused the bug and is in the fix Alen made a few comments earlier.

**Comment by andrewcraik** [comment]
ok great thanks - just confirming there wasn't a trees/common issue - looks to be codegen - thanks!

**Comment by AlenBadel** [comment]
In regards to the crash on power, the following are the trees generated.
```
 n489n    (  0)  treetop                                                                              [    0x7fffcea98890] bci=[6,6,1361] rc=0 vc=19 vn=- li=45 udi=- nc=1
 n488n    (  0)    call  jdk/internal/misc/Unsafe.setMemory(Ljava/lang/Object;JJB)V[#487  final virtual Method -720] [flags 0x20500 0x0 ] ()  [    0x7fffcea98840] bci=[6,6,1361] rc=0 vc=19 vn=- li=45 udi=- nc=5 flg=0x20
 n483n    (  0)      aload  <temp slot 6>[#477  Auto] [flags 0x20000007 0x0 ] (X!=0 sharedMemory )    [    0x7fffcea986b0] bci=[6,0,1361] rc=0 vc=19 vn=- li=45 udi=- nc=0 flg=0x4
 n484n    (  0)      aconst NULL (in GPR_0641)                                                        [    0x7fffcea98700] bci=[6,1,1361] rc=0 vc=19 vn=- li=45 udi=36048 nc=0
 n485n    (  0)      lload  buf<auto slot 6>[#392  Auto] [flags 0x4 0x0 ] (in GPR_0642)               [    0x7fffcea98750] bci=[6,2,1361] rc=0 vc=19 vn=- li=45 udi=36320 nc=0
 n486n    (  0)      lload  alignedSize<auto slot 4>[#386  Auto] [flags 0x4 0x0 ] (in GPR_0644)       [    0x7fffcea987a0] bci=[6,3,1361] rc=0 vc=19 vn=- li=45 udi=36976 nc=0
 n495n    (  0)      iconst 0 (in GPR_0646)                                                           [    0x7fffcea98a70] bci=[-1,41,93] rc=0 vc=19 vn=- li=45 udi=37632 nc=0
```

I'll be taking a second look at the generated sequence to see if there were any similar assumptions, otherwise we should be ready for a PR. 

On the XlinuxXL front, I'm checking if there are similar assumptions taken (I.e size of data treated as an int), otherwise there's very little I can do without being able to reproduce the failure.

**Comment by andrewcraik** [comment]
so the trees do look correct so the issue would just be incorrect widths in the codegen - thanks for the great analysis @AlenBadel

**Comment by gita-omr** [comment]
I think we still need to understand the whole flow: which opcodes and evaluators are involved. For example, we might be reusing some code that normally accepts int length.

**Comment by AlenBadel** [comment]
> I think we still need to understand the whole flow: which opcodes and evaluators are involved. For example, we might be reusing some code that normally accepts int length.

I agree. I will have an update soon with the summary of the issue from a top-down approach.

**Comment by keithc-ca** [comment]
It occurs to me that this would have been avoided if we had some kind of 'compiler' to check uses like that, along the lines of the error you get in Java passing a long to a function that expects an int (e.g. `Integer.valueOf(42L)`). To make it concrete here, we would have a class whose job is to form an instruction stream, that class would have methods `srawi(Register32,Register32,int32_t)` and  `sradi(Register64,Register64,int64_t)` - calling the former with 64-bit registers or a a 64-bit value would not compile.

**Comment by gita-omr** [comment]
Yeah, it's something to consider in the future. Basically, Power codegen's assumption is that a register containing an Int can have an undefined value in it's upper half. So Int and Long types need to be treated differently. However we only have Register class and not Register32 and Register64.

**Comment by gita-omr** [comment]
BTW : I think we need to remove the Intel  label and assume that this issue  will only take care of the Power failure.

**Comment by pshipton** [comment]
Removed the intrel label, this platform should be covered by https://github.com/eclipse/openj9/issues/10580

**Comment by AlenBadel** [comment]
> `setmemoryEvaluator `is defined in OMR without much description of the expected input parameters [1]

This description can certainly be added. 

> Note that the length information is conveyed via creating an arrayset node that normally takes Int length.

This is true, an ArraySet does normally allocate an array length of type int due to indexing limitations. The limitation is historical, and based off the class ArraySet inherits; ArrayList which itself is an implementation of List.  

Looking through the remainder of the Unsafe methods, we can see that we tend use the underlying array evaluators. From what I've seen they all support a length which could be long, or int.

For example, such case as `Unsafe.CopyMemory`[1] which uses the underlying `arrayCopyEvaluator`[2]. We see that the underlying arrayCopyEvaluator does support sizes of long[3].

I believe as long as the evaluators we use support the initialization of a contiguous memory segment which has a length argument of type long then we could easily justify it's use to compile and evaluate Unsafe methods. 
To note, arrays do not have a bound on how large they can be, rather they have a bound on the number of objects (ArrayList based implementations i.e ArraySet) or primitives they index (Integer.MAX_VALUE). Hence the process of initialization of contiguous memory is identical in the case of Unsafe.CopyMemory, or Unsafe.SetMemory compared to arrays themselves.

> I am not sure we can change OMR method to adjust it to a particular consumer, openj9, especially since even this consumer passes information in somewhat inconsistent way.

The evaluator should support both long, or int length values. This is consistent with many of our other evaluators. Java is not the only language that has the capability to allocate, and initialize a very-large contiguous region of memory.

[1] https://github.com/eclipse/omr/blob/0b9653c23ba22247f66c11ad6df8a3b46c2c4738/compiler/p/codegen/OMRTreeEvaluator.cpp#L4276-L4277
[2] https://github.com/eclipse/openj9/blob/9369e4e005a130411ae2606a25d08588a0920c7a/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L12842
[3] https://github.com/eclipse/openj9/blob/9369e4e005a130411ae2606a25d08588a0920c7a/runtime/compiler/p/codegen/J9TreeEvaluator.cpp#L12890-L12891

**Comment by gita-omr** [comment]
`The evaluator should support both long, or int length values.`

Do you mean we should be able to pass nodes with the length child being both Int and Long type?

**Comment by AlenBadel** [comment]
> Do you mean we should be able to pass nodes with the length child being both Int and Long type?

They can be either Long, or Int. We currently don't call `setMemoryEvaluator` for any other case other than `Unsafe.SetMemory()` which accepts a long a long length value.

However, it's better illustrated with the example of `arrayCopyEvaluator`. 
Which is used by `Unsafe.CopyMemory` as it accepts a length value of long type, and at the same time is used with `System.arraycopy` which accepts a length of int type.

**Comment by keithc-ca** [comment]
I had a look at how `writeCrashDataToConsole()` might fail, but didn't see anything obvious. Given that the current theory is that the generated code was clearing memory that should have been left alone, it's reasonable that some of the data examined by `writeCrashDataToConsole()` was corrupted and so a second fault should not be unexpected.

**Comment by AlenBadel** [comment]
OMR PR: https://github.com/eclipse/omr/pull/5590

**Comment by AlenBadel** [comment]
The above PR has been merged. 

Fix will be merged into R0.23 as well. 
https://github.com/eclipse/openj9-omr/pull/81

**Comment by gita-omr** [comment]
@andrew-m-leonard  can this be closed?

**Comment by pshipton** [comment]
I'll go ahead and close it since the fix has been merged, we can open a new issue if another problem is found.
